### PR TITLE
Add ldap group/membership tests

### PR DESCRIPTION
**Pull request by arif-ali**
Adds a test to check for groups that are coming from LDAP.

Adds a test to ensure that openstack is able to check the membership
of a user in the group.

Signed-off-by: Arif Ali <arif.ali@canonical.com>

### CODE REVIEW

**Comment by arif-ali** [comment]
This passed with `func-smoke` for keystone-ldap charm, but now running full func tests, we get the following error

```
2020-10-08 08:40:25 [INFO] keystoneauth1.exceptions.http.Forbidden: You are not authorized to perform the requested action: identity:list_groups_for_user (HTTP 403) (Request-ID: req-f9e105d0-194f-4b1e-b91a-edb82d1e0540)
```
This is from the test test_102_keystone_ldap_group_membership, maybe requires extra policy changes

**Comment by arif-ali** [comment]
This will conflict and add on to of #458

**Comment by arif-ali** [comment]
func-smoke: https://paste.ubuntu.com/p/HwdQjqdmc3/

after making the following modifications

```diff
diff --git a/src/test-requirements.txt b/src/test-requirements.txt
index d3c9be8..3bd6c04 100644
--- a/src/test-requirements.txt
+++ b/src/test-requirements.txt
@@ -5,4 +5,4 @@
 #
 # Functional Test Requirements (let Zaza's dependencies solve all dependencies here!)
 git+https://github.com/openstack-charmers/zaza.git#egg=zaza
-git+https://github.com/openstack-charmers/zaza-openstack-tests.git#egg=zaza.openstack
+git+https://github.com/arif-ali/zaza-openstack-tests.git@ldap-groups-membership-tests#egg=zaza.openstack
```

File: zaza/openstack/charm_tests/keystone/tests.py [hunk 1]
```diff
@@ -436,6 +444,62 @@ def _find_keystone_v3_user(self, username, domain):
         )
         return None
 
+    def _find_keystone_v3_group(self, group, domain):
+        """Find a group within a specified keystone v3 domain.
+
+        :param str group: Group to search for in keystone
+        :param str domain: username selected from which domain
```
Comment (AurelienLourot) [inline zaza/openstack/charm_tests/keystone/tests.py:451]: s/username/group/

File: zaza/openstack/charm_tests/keystone/tests.py [hunk 2]
```diff
@@ -436,6 +444,62 @@ def _find_keystone_v3_user(self, username, domain):
...
+                if searched_group.name.lower() == group.lower():
+                    return searched_group
+
+        logging.debug(
+            "Group {} was not found. Returning None.".format(group)
+        )
+        return None
+
+    def _find_keystone_v3_user_in_group(self, user, group, domain):
+        """Find a user in group group within a specified keystone v3 domain.
```
Comment (AurelienLourot) [inline zaza/openstack/charm_tests/keystone/tests.py:475]: s/group group/a group/

File: zaza/openstack/charm_tests/keystone/tests.py [hunk 3]
```diff
@@ -436,6 +444,62 @@ def _find_keystone_v3_user(self, username, domain):
...
+
+        :param str user: User to search for in keystone
+        :param str group: Group to search for in keystone
+        :param str domain: username/group selected from which domain
+        :return: return group if found
```
Comment (AurelienLourot) [inline zaza/openstack/charm_tests/keystone/tests.py:480]: s/group/user/

File: zaza/openstack/charm_tests/keystone/tests.py [hunk 4]
```diff
@@ -436,6 +444,62 @@ def _find_keystone_v3_user(self, username, domain):
...
+        """
+        for ip in self.keystone_ips:
+            logging.info('Keystone IP {}'.format(ip))
+            session = openstack_utils.get_keystone_session(
+                openstack_utils.get_overcloud_auth(address=ip))
+            client = openstack_utils.get_keystone_session_client(session)
+
+            domain_users = client.users.list(
+                domain=client.domains.find(name=domain).id,
+                group=self._find_keystone_v3_group(group, domain).id,
```
Comment (AurelienLourot) [inline zaza/openstack/charm_tests/keystone/tests.py:491]: Apart from this line it's the same implementation as `_find_keystone_v3_user()`. Couldn't we have extended that function with an optional `group` argument instead?

**Review by AurelienLourot — CHANGES_REQUESTED** [review]
LGTM in general, just a few nitpicks, thanks!

File: zaza/openstack/charm_tests/keystone/tests.py (hunk 4 above)
Comment (arif-ali) [reply zaza/openstack/charm_tests/keystone/tests.py:491]: hmm, nice catch, thanks, I will sort it, and push a new update later.

I did think at one point, that there was some repetitiveness, but couldn't fathom it :)

**Review by AurelienLourot — APPROVED** [review]

**Pull request merged**